libsolidity/lsp/LanguageServer.cpp

```diff
@@ -0,0 +1,955 @@
+	{
+		// For scope members, jump to the naming symbol of the referencing declaration of this member.
+		auto const declaration = n->annotation().referencedDeclaration;
+		auto const location = declarationPosition(declaration);
+		if (location.has_value())
+			locations.emplace_back(location.value());
+		else
+			trace("gotoDefinition: declaration not found.");
+	}
+	else if (auto const sourceIdentifier = dynamic_cast<Identifier const*>(sourceNode))
+	{
+		// For identifiers, jump to the naming symbol of the definition of this identifier.
+		if (Declaration const* decl = sourceIdentifier->annotation().referencedDeclaration)
+			if (auto location = declarationPosition(decl); location.has_value())
+				locations.emplace_back(move(location.value()));
+		// if (auto location = declarationPosition(sourceIdentifier->annotation().referencedDeclaration); location.has_value())
+		// 	locations.emplace_back(move(location.value()));
+
+		for (auto const declaration: sourceIdentifier->annotation().candidateDeclarations)
+			if (auto location = declarationPosition(declaration); location.has_value())
+				locations.emplace_back(move(location.value()));
+	}
+	else
+		trace("gotoDefinition: Symbol is not an identifier. "s + typeid(*sourceNode).name());
+
+	Json::Value reply = Json::arrayValue;
+	for (SourceLocation const& location: locations)
+		reply.append(toJson(m_basePath, location));
+	m_client->reply(_id, reply);
+}
+
+void LanguageServer::handle_textDocument_highlight(MessageId _id, Json::Value const& _args)
+{
+	auto const dpos = extractDocumentPosition(_args);
+
+	Json::Value jsonReply = Json::arrayValue;
+	for (DocumentHighlight const& highlight: semanticHighlight(dpos))
+	{
+		Json::Value item = Json::objectValue;
+		item["range"] = toJsonRange(highlight.location);
+		if (highlight.kind != DocumentHighlightKind::Unspecified)
+			item["kind"] = static_cast<int>(highlight.kind);
+
+		jsonReply.append(item);
+	}
+	m_client->reply(_id, jsonReply);
+}
+
+void LanguageServer::handle_textDocument_references(MessageId _id, Json::Value const& _args)
+{
+	auto const dpos = extractDocumentPosition(_args);
+
+	trace(
+		"find all references: " +
+		dpos.path + ":" +
+		to_string(dpos.position.line) + ":" +
+		to_string(dpos.position.column)
+	);
+
+	auto file = m_fileReader->sourceCodes().find(dpos.path);
+	if (file == m_fileReader->sourceCodes().end())
+	{
+		Json::Value emptyResponse = Json::arrayValue;
+		m_client->reply(_id, emptyResponse); // reply with "No references".
+		return;
+	}
+
+	if (!m_compilerStack)
+		compile(dpos.path);
+	solAssert(m_compilerStack.get() != nullptr, "");
+
+	auto const sourceNode = findASTNode(dpos.position, dpos.path);
+	if (!sourceNode)
+	{
+		Json::Value emptyResponse = Json::arrayValue;
+		m_client->reply(_id, emptyResponse); // reply with "No references".
+		return;
+	}
+
+	auto locations = vector<SourceLocation>{};
+	if (auto const sourceIdentifier = dynamic_cast<Identifier const*>(sourceNode))
```

Comment:
What is the difference between this piece of code and the one in LanguageServer::references ?